031-DevOpsWithGitHub/Coach/Setup/wth-setup.sh

```diff
@@ -0,0 +1,117 @@
+#   - You have a directory containing the Hackathon resource files (set the name as a environment variable below before running)
+#   - You have a password for the users and coaches (set the name as a environment variable below before running)
+#
+# Interesting Fact: Copilot wrote a massive percentage of this script and this desciption :-)
+#
+###################################################################################################################################
+
+# The number of groups in the hackathon
+GROUP_COUNT=25
+
+# The Azure subscription ID
+SUBSCRIPTION_ID=<AZURE-SUBSCRIPTION-ID>
+
+# The organisation in GitHub
+ORGANISATION=<GITHUB-ORGANISATION-NAME e.g. wth-devops-london>
+
+# The directory containing the Hackathon resource filed
+RESOURCE_DIR=./Resources
+
+# The password for the users and coaches
+PASSWORD=
+
+
+# Set the subscription
+az account set --subscription $SUBSCRIPTION_ID
+DOMAIN=$(az rest --method get --url 'https://graph.microsoft.com/v1.0/domains?$select=id' --query '"value"' | jq '.[0].id' --raw-output) 
+
+mkdir -p AzureCredentials
+# Create all the Azure resource groups and user/service principals
+for i in $(seq $GROUP_COUNT); do
+    echo "Creating resource group wth-team$i-rg"
+    az group create --name wth-team$i-rg --location uksouth
+    
+    COACH_UPN=wth-coach-team$i@$DOMAIN
+    TEAM_UPN=wth-team$i@$DOMAIN
+
+    echo "Creating coach user $COACH_UPN"
+    az ad user create --display-name wth-team$i-coach --password $PASSWORD --user-principal-name $COACH_UPN
+
+    echo "Creating team user $TEAM_UPN"
+    az ad user create --display-name wth-team$i-user  --password $PASSWORD --user-principal-name $TEAM_UPN
+
+    echo "Creating role assignment $COACH_UPN"
+    az role assignment create --assignee $COACH_UPN  --role contributor --resource-group wth-team$i-rg
+
+    echo "Creating role assignment $TEAM_UPN"
+    az role assignment create --assignee $TEAM_UPN  --role contributor --resource-group wth-team$i-rg
+
+    echo "Creating service principal into ./AzureCredentials/principal-team$i.json"
+    az ad sp create-for-rbac --name "wth-team$i" --role contributor \
+                        --scopes /subscriptions/$SUBSCRIPTION_ID/resourceGroups/wth-team$i-rg \
+                        --sdk-auth > ./AzureCredentials/principal-team$i.json
+done
+
+
+# Make some directories to support the process and create one as a git repo
+mkdir -p ResourcesRepo
+git init ./ResourcesRepo
+
+# Copy the files for the team repo into local git repo
+cp -r $RESOURCE_DIR/ ./ResourcesRepo
+cd ./ResourcesRepo
+git add .
+git commit -m "Initial commit"
+
+# Create the team repos
+for i in $(seq $GROUP_COUNT); do
+    # Create a team repo
+    echo "Creating team repo $ORGANISATION/wth-team$i"
+    gh repo create $ORGANISATION/wth-team$i --public --confirm
+
+    echo "Add the team$i repo as a remote and push the resources"
+    git remote add team$i https://github.com/$ORGANISATION/wth-team$i.git
+    git push team$i main
```

Comment:
I am getting an error here unless I set it to master instead of main. OR I create the branch main first.  Let me know if im doing something wrong here.

Reply:
One additional thing I noticed, since we are storing credentials for this training event in a separate repo/issue.  Do we want to have a tear down script or provide some guidance to eliminate some of these credentials laying around afterwards?

Reply:
I make an assumption that everyone has switch their default branch now from master to main. Changing the script to factor in those who have yet to do this by changing the branch name to main. This will error for those of us who have main as default but as a no-op.

On the clean-up script, agreed I will write such a script at a future point to tidy up the AAD. In our case a simple deletion of the Resource Groups and the GitHub organisation will mean these users have role assignments to nothing and the credentials in their issue form will be no more.